tools/color-shapes/merge.ts

```diff
@@ -0,0 +1,58 @@
+/* eslint-disable @typescript-eslint/explicit-function-return-type */
+import { Bilateral3DMap } from "./bilateral-3d-maps.js"
+import { DataPoint } from "./data-point.js"
+import { Vec3Like } from "./vec3.js"
+
+export function collapseVertices<T>(
+  intTriangles: DataPoint<T | null>[],
+  collapse: (data: T[]) => T | null
+) {
+  let map = new Bilateral3DMap<number>()
+  let dataMap = new Bilateral3DMap<T[]>()
+  let indices: number[] = []
+
+  let weldedVerts: Vec3Like[] = []
+  let weldedData: (T | null)[] = []
+
+  for (let i = 0; i < intTriangles.length; i += 3) {
+    let a = intTriangles[i + 0]
+    let b = intTriangles[i + 1]
+    let c = intTriangles[i + 2];
+
+    ([a,b,c]).forEach(point => {
+      let {x,y,z} = point.pos
+      let data = point.data
+
+      // Set data`s
+      let datas = dataMap.getValue(x,y,z) || []
+      if (data) datas.push(data)
+      dataMap.set(datas, x,y,z)
+
+      if (!map.has(x, y, z)) {
+        // Setp verts
+        map.set(map.valuesCount, x, y, z)
+
+        // Generate welded verts
+        let [[wx, wy, wz]] = map.getKeys(map.valuesCount - 1)
+        weldedVerts.push({x: wx, y: wy, z: wz})
+
+        // Generate datas
+        let items = dataMap.getValue(wx, wy, wz)
+        let collapsedItem = collapse(items || [])
+        weldedData.push(collapsedItem)
+      }
+
+      // Generate indices
+      let index = map.getValue(x, y, z)
+      if (index !== undefined) {
+        indices.push(index)
+      }
+    })
+  }
+
+  return {
+    vertices: weldedVerts,
```

Comment:
The spaces is about 6 MB. It is a lot!

I opened a file and found plenty space for optimization:

```
{"vertices":[0.01746031746031746,0.013886243386243386,0.010551587301587303,0.01210952380952381,0.014587301587301588,0.013103174603174603,0.014944444444444446,0.010175925925925927,0.016351851851851854,0.010324074074074074,0.013324074074074073,0.019083333333333334,0.014055555555555552,0.00887037037037037,0.024537037037037038,0.009712962962962963,0.012861111111111108,0.027962962962962964,0.01388888888888889,0.008541666666666666,0.03409722222222222,0.00962962962962963,0.012777777777777775,0.03777777777777778,0.01388888888888889,0.008518518518518517,0.04407407407407407,0.00962962962962963,0.012777777777777775,0.04777777777777778,0.01388888888888889,0.008518518518518517,0.05407407407407408,0.00962962962962963,0.012777777777777775,0.05777777777777778,0.01388888888888889,0.008518518518518517,0.06407407407407407,
```

We do not need this precision. Just by rounding the numbers we can reduce the size a lot.